### PR TITLE
tls: introduce Verify() helper and ClientHelloInfo alias

### PR DESCRIPTION
**Pull request by amery**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new type alias for easier reference to `tls.ClientHelloInfo`.
	- Added comprehensive functions for verifying TLS certificates, enhancing security checks and error handling.

- **Bug Fixes**
	- Improved error handling during certificate validation to ensure robustness.

### CODE REVIEW

File: tls/verify.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	case cert.Leaf == nil || len(cert.Certificate) == 0:
+		return &x509utils.ErrInvalidCert{
+			Reason: "missing leaf certificate",
+		}
+	case len(cert.Leaf.Raw) == 0,
+		cert.Leaf.NotAfter.After(time.Now()), cert.Leaf.NotBefore.Before(time.Now()),
+		!bytes.Equal(cert.Leaf.Raw, cert.Certificate[0]):
+		return &x509utils.ErrInvalidCert{
+			Reason: "invalid leaf certificate",
+		}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go:32]: good catch

File: tls/verify.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func doVerifyRoots(cert *tls.Certificate, roots *x509.CertPool) error {
+	opt := x509.VerifyOptions{
+		Roots:         roots,
+		Intermediates: x509.NewCertPool(),
+	}
+	for _, b := range cert.Certificate[1:] {
+		opt.Intermediates.AppendCertsFromPEM(b)
+	}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go:55]: I completely forgot those are DER

**Comment by amery** [comment]
amended mistakes caught by @coderabbitai

File: tls/std.go [hunk 3]
```diff
@@ -7,6 +7,9 @@ type (
 	// Certificate is an alias of the standard [tls.Certificate]
 	Certificate = tls.Certificate
 
+	// ClientHelloInfo is an alias of the standard [tls.ClientHelloInfo].
+	ClientHelloInfo = tls.ClientHelloInfo
+
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/std.go:12]: would that affect the use of the Store interface?

File: tls/verify.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+			count++
+		}
+		return e2 == nil
+	})
+
+	if e1 != nil {
+		return e1
+	}
+	return e2
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go]: it should be gone now

File: tls/verify.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go]: it should be fixed now

File: tls/verify.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go]: it should be fixed now

File: tls/verify.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	case !x509utils.ValidCertKeyPair(cert.Leaf, cert.PrivateKey):
+		return &x509utils.ErrInvalidCert{
+			Reason: "invalid private key",
+		}
+	case roots != nil:
+		return doVerifyRoots(cert, roots)
+	default:
+		return doVerifyNoRoots(cert)
+	}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go]: what metric do you use to determine a list of if conditions is more readable that the switch?

File: tls/verify.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		return &x509utils.ErrInvalidCert{
+			Reason: "none provided",
+		}
+	case cert.Leaf == nil || len(cert.Certificate) == 0:
+		return &x509utils.ErrInvalidCert{
+			Reason: "missing leaf certificate",
+		}
+	case len(cert.Leaf.Raw) == 0,
+		cert.Leaf.NotAfter.Before(time.Now()),
+		cert.Leaf.NotBefore.After(time.Now()),
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go]: refactored

File: tls/verify.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	for i, der := range cert.Certificate[1:] {
+		c, err := doVerifyDERCert(i+1, der)
+		if err != nil {
+			return core.Wrap(err, "ReadCertificate")
+		}
+
+		opt.Intermediates.AddCert(c)
+	}
+	_, err := cert.Leaf.Verify(opt)
+	return err
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/verify.go]: this should be resolved now

**Comment by amery** [comment]
> **Actionable comments posted: 5**
> 
> 🧹 Outside diff range and nitpick comments (1)
> > tls/std.go (1)> `10-12`: **LGTM! Consider a minor comment style improvement.**
> > > The addition of the `ClientHelloInfo` type alias is consistent with the existing pattern in the file and maintains the alphabetical order of the aliases. This change enhances the package's usability by providing a convenient alias for `tls.ClientHelloInfo`.
> > > For consistency with the other comments, consider removing the period at the end of the comment:
> > > ```diff
> > > - // ClientHelloInfo is an alias of the standard [tls.ClientHelloInfo].
> > > + // ClientHelloInfo is an alias of the standard [tls.ClientHelloInfo]
> > > ```
> 
> 📜 Review details

we prefer them with a period at the end of each sentence.

**Review by karasz — APPROVED** [review]
+1

**Pull request merged**